### ISSUE TITLE
Add custom fields for catalog objects (i.e. plans, products)

### ISSUE THREAD

**Issue by bvisonl** [killbill/killbill#1938]
I've been trying to figure out how could I get custom fields on all the entities that are inside the catalog. I've spent quite some time digging around and I think that I get the "why" custom fields are not allowed in Catalog, Product, Plans, etc. However, there may be a lot of use for them. 

An easy way out of this that I can see is just updating the schema to allow a section of key/value items that would later be used by the `kbcli`. I'm thinking that, for example, the `Plan` entity could have a field called `<CustomFields>...</CustomFields>` which could end up being just a `map[string]string` or something like that in the `kbcli`. That would make it easy enough so that it doesn't break any existing logic related to the catalog and flexible enough so that anyone using the `kbcli` can extend the functionality of KillBill even more without needing to write a whole plugin. 

I think that the schema is so closed and tight in validation that it makes it a little bit hard to build logic around anything catalog-related that's outside the rules / alignments scope.

Right now I am attempting to enable this by modifying the `xsd` but that doesn't seem to be working when I try to upload the catalog. 

### Example of what I want in the XSD
```xml
    <xs:complexType name="defaultPlan">
        <xs:complexContent>
            <xs:extension base="validatingConfig">
                <xs:sequence>
                    <xs:element minOccurs="0" name="effectiveDateForExistingSubscriptions" type="xs:dateTime"/>
                    <xs:element name="product" type="xs:IDREF"/>
                    <xs:element minOccurs="0" name="recurringBillingMode" type="billingMode"/>
                    <xs:element minOccurs="0" name="initialPhases">
                        <xs:complexType>
                            <xs:sequence>
                                <xs:element maxOccurs="unbounded" minOccurs="0" name="phase" type="defaultPlanPhase"/>
                            </xs:sequence>
                        </xs:complexType>
                    </xs:element>

                   <!-- Note this here -->
                    <xs:element minOccurs="0" name="customFields">
                        <xs:complexType>
                            <xs:sequence>
                                <xs:element maxOccurs="unbounded" minOccurs="0" name="customField" type="defaultCustomField"/>
                            </xs:sequence>
                        </xs:complexType>
                    </xs:element>

                    <xs:element name="finalPhase" type="defaultPlanPhase"/>
                    <xs:element minOccurs="0" name="plansAllowedInBundle" type="xs:int"/>
                </xs:sequence>
                <xs:attribute name="name" type="xs:ID" use="required"/>
                <xs:attribute name="prettyName" type="xs:string"/>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
   
   ...

    <xs:complexType name="defaultCustomField">
        <xs:complexContent>
            <xs:extension base="validatingConfig">
                <xs:attribute name="key" type="xs:string" use="required"/>
                <xs:attribute name="value" type="xs:string" use="required"/>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
```

So that later in the catalog I can do:

```xml
...
        <plans>
            <plan name="my-plan">
                <product>THE_PRODUCT</product>
                <initialPhases>
                </initialPhases>
                <customFields>
                    <customField key="foo" value="bar" />
                </customFields>
...
```

Any idea on how to get this done is welcome!

**Comment by kakarot-101** [comment]
Absolutely! Incorporating custom fields for catalog objects, such as plans and products, enables tailored support for diverse information needs. By adding custom fields, you can enhance the catalog's versatility and accommodate various requirements, ensuring it can efficiently support your business operations. Whether it's capturing unique identifiers, specialized attributes, or specific categorizations, these custom fields empower your catalog to provide comprehensive support for your organization's evolving needs. Let me know if you need further guidance on implementing this enhanced [support](https://fescoebill.pk/) functionality!

**Comment by amjad0989** [comment]
Of course! Custom fields for catalog objects, such products and plans, allow customized support for many types of information requirements. You can increase the catalog's adaptability and meet different needs by adding custom fields, so you can be sure it will effectively assist your business operations. These custom fields enable your catalog to offer comprehensive support for the changing needs of your company, whether it be through the capture of unique identifiers, specialty qualities, or particular categorizations. Please let me know if you require any additional help click here https://myvipbio.com/instagram-bio-for-girls-attitude/

**Comment by shaun-forgie** [comment]
These sorts of plan type meta data fields defined in the catalog would then be added to subscriptions which in turn need to be included on the documents [invoices] that are generated during billing cycles. 

We are currently discussing a related invoicing meta data feature here https://github.com/killbill/killbill/issues/2074

**Comment by shaun-forgie** [comment]
@kakarot-101 can you get in touch with me to discuss this.

**Comment by shaun-forgie** [comment]
The key requirement the Bill Rush team has is being able to setup and configure the information that is included on invoices. Conceptually the composition relationship that exists between elements in the model is Accounts -> subscriptions -> invoices. Meta data included on parent subscriptions and accounts become available for for stamping / inclusion on invoices. This is important for analysing this data after the invoices are generated and delivered and is particularly useful for revenue recognition and producing account journal entries for down steam finance systems. 

By specifying meta-data requirements in the catalog at the product or plan level this essentially allows for these fields with a default values to be copied across into the subscription during creation. As discussed above this then allows these fields to be included in the  "invoice meta data stamping" by way of configuration file.
